### PR TITLE
GMOS fluxing

### PR DESCRIPTION
**Pull request by profxj**
A few bits and pieces related to fluxing and GMOS.

And this PR now requires RA/DEC be floats within PypeIt.
And this will propogate to the PypeIt file.

### CODE REVIEW

**Comment by profxj** [comment]
============================ test session starts =============================
platform linux -- Python 3.7.3, pytest-5.3.5, py-1.8.1, pluggy-0.13.0
rootdir: /data/Projects/Python/PypeIt
plugins: cov-2.7.1, hypothesis-5.6.0, remotedata-0.3.1, astropy-header-0.1.2, openfiles-0.4.0, arraydiff-0.3, doctestplus-0.4.0
collected 197 items                                                           

pypeit/tests/test_arc.py .                                              [  0%]
pypeit/tests/test_biasframe.py ....                                     [  2%]
pypeit/tests/test_bpmimage.py ..                                        [  3%]
pypeit/tests/test_calibrationimage.py .                                 [  4%]
pypeit/tests/test_calibrations.py ...........                           [  9%]
pypeit/tests/test_coadd.py .                                            [ 10%]
pypeit/tests/test_cooked.py .                                           [ 10%]
pypeit/tests/test_coremeta.py .                                         [ 11%]
pypeit/tests/test_datacontainer.py ......                               [ 14%]
pypeit/tests/test_edgetrace.py .....                                    [ 16%]
pypeit/tests/test_flatfield.py .                                        [ 17%]
pypeit/tests/test_flexure.py .                                          [ 17%]
pypeit/tests/test_flux.py ....                                          [ 19%]
pypeit/tests/test_fluxspec.py ..                                        [ 20%]
pypeit/tests/test_load.py .                                             [ 21%]
pypeit/tests/test_load_images.py ...........                            [ 26%]
pypeit/tests/test_masterframe.py .                                      [ 27%]
pypeit/tests/test_match.py ....                                         [ 29%]
pypeit/tests/test_metadata.py .....                                     [ 31%]
pypeit/tests/test_moment.py ......                                      [ 35%]
pypeit/tests/test_msgs.py ..                                            [ 36%]
pypeit/tests/test_opticalmodel.py ....                                  [ 38%]
pypeit/tests/test_parse.py ...                                          [ 39%]
pypeit/tests/test_pca.py ....                                           [ 41%]
pypeit/tests/test_processimage.py ..                                    [ 42%]
pypeit/tests/test_procimg.py .                                          [ 43%]
pypeit/tests/test_pydl.py .                                             [ 43%]
pypeit/tests/test_pypeit.py ..                                          [ 44%]
pypeit/tests/test_pypeitimage.py ....                                   [ 46%]
pypeit/tests/test_pypeitpar.py .................................        [ 63%]
pypeit/tests/test_pypeitsetup.py .......                                [ 67%]
pypeit/tests/test_qa.py .                                               [ 67%]
pypeit/tests/test_save.py .                                             [ 68%]
pypeit/tests/test_scienceimage.py ....                                  [ 70%]
pypeit/tests/test_scripts.py .........                                  [ 74%]
pypeit/tests/test_setups.py ...............                             [ 82%]
pypeit/tests/test_slitmask.py .                                         [ 82%]
pypeit/tests/test_specobj.py ....                                       [ 84%]
pypeit/tests/test_specobjs.py ....                                      [ 86%]
pypeit/tests/test_spectrographs.py ............                         [ 92%]
pypeit/tests/test_traceimage.py ..                                      [ 93%]
pypeit/tests/test_utils.py ....                                         [ 95%]
pypeit/tests/test_vcorr.py ..                                           [ 96%]
pypeit/tests/test_wavecalib.py .                                        [ 97%]
pypeit/tests/test_waveimage.py .                                        [ 97%]
pypeit/tests/test_wavemodel.py .                                        [ 98%]
pypeit/tests/test_wavetilts.py ...                                      [100%]

================ 197 passed, 440 warnings in 764.73s (0:12:44) ================

File: pypeit/core/wavecal/templates.py [hunk 1]
```diff
@@ -434,7 +438,20 @@ def main(flg):
         wfile = os.path.join(template_path, 'Keck_LRIS', 'R600_5000', 'MasterWaveCalib_B_1_01.json')
         build_template(wfile, slits, lcut, binspec, outroot, lowredux=False)
 
-    if flg & (2**13):  # Magellan/MagE
+    if flg & (2**27):  # R600/7500
```
Comment (kbwestfall) [inline pypeit/core/wavecal/templates.py]: See #850 , not necessarily for this PR.

File: pypeit/io.py [hunk 2]
```diff
@@ -250,7 +252,10 @@ def initialize_header(hdr=None):
     # the versions of all packages included in the requirements.txt
     # file?
     if hdr is None:
-        hdr = fits.Header()
+        if primary:
```
Comment (kbwestfall) [inline pypeit/io.py]: Why is this necessary?  I played with this a bit and I may understand why you did this, but did you see some behavior where astropy wasn't treating primary vs non-primary headers correctly when the `HDUList` was instantiated?

File: pypeit/metadata.py [hunk 3]
```diff
@@ -174,7 +174,12 @@ def _build(self, files, strict=True, usrdata=None):
         # Build the table
         for idx, ifile in enumerate(_files):
             # User data (for frame type)
-            usr_row = None if usrdata is None else usrdata[idx]
+            if usrdata is None:
+                usr_row = None
+            else:
+                # Check
+                assert os.path.basename(ifile) == usrdata['filename'][idx]
```
Comment (kbwestfall) [inline pypeit/metadata.py]: Another pet peeve:  We shouldn't be putting assertions in production-level code.  They're good for checking that the coding is right, but once it's tested, these should either be removed or replaced with an exception that raises a meaningful error.

File: pypeit/metadata.py [hunk 4]
```diff
@@ -359,10 +364,11 @@ def merge(self, usrdata, match_type=True):
 
         # Convert types if possible
         existing_keys = list(set(self.table.keys()) & set(usrdata.keys()))
+        radec_done = False
         if len(existing_keys) > 0 and match_type:
             for key in existing_keys:
                 if len(self.table[key].shape) > 1:  # NOT ALLOWED!!
-                    debugger.set_trace()
+                    import pdb; pdb.set_trace()
```
Comment (kbwestfall) [inline pypeit/metadata.py]: Replace with embed()?  Or (preferably) remove?

File: pypeit/metadata.py [hunk 5]
```diff
@@ -371,7 +377,18 @@ def merge(self, usrdata, match_type=True):
                 nones = usrdata[key] == 'None'
                 usrdata[key][nones] = None
                 # Rest
-                usrdata[key][~nones] = usrdata[key][~nones].astype(dtype)
+                # Allow for str RA, DEC (backwards compatability)
+                if key in ['ra', 'dec']:
```
Comment (kbwestfall) [inline pypeit/metadata.py]: Why not:
```
if key in ['ra', 'dec'] and not radec_done:
    ras, decs = meta.convert_radec(...)
    ...
else:
    usrdata[key][~nones] = usrdata[key][~nones].astype(dtype)
```

File: pypeit/scripts/show_2dspec.py [hunk 6]
```diff
@@ -133,7 +133,15 @@ def main(args):
     slit_ids = [edgetrace.get_slitid(shape, tslits_dict['slit_left'], tslits_dict['slit_righ'],
                                      ii)[0] for ii in range(tslits_dict['slit_left'].shape[1])]
     # Show the bitmask?
-    mask_in = mask if args.showmask else None
+    if args.showmask:
+        mask_in = mask
+        # Unpack the bitmask
+        bitMask = ImageBitMask()
+        bpm, crmask, satmask, minmask, offslitmask, nanmask, ivar0mask, ivarnanmask, extractmask \
```
Comment (kbwestfall) [inline pypeit/scripts/show_2dspec.py]: This unpack is not needed, right?

File: pypeit/sensfunc.py [hunk 7]
```diff
@@ -255,12 +261,25 @@ def splice(self, wave):
...
+            zero_values = interp_func(wave_splice[zeros])
+            sensfunc_splice[zeros] = zero_values
+
         self.steps.append(inspect.stack()[0][3])
 
         return wave_splice, sensfunc_splice
 
     def show(self):
-        pass
+        plt.figure()
```
Comment (kbwestfall) [inline pypeit/sensfunc.py]: Do you need the `plt.figure()` and `plt.close()` lines?

File: pypeit/specobjs.py [hunk 8]
```diff
@@ -506,23 +517,35 @@ def __repr__(self):
...
 
-    def build_header(self, head_fitstbl, spectrograph):
+    def build_header(self, head_fitstbl, head2d, spectrograph):
         """
         Builds the spec1d file header from the fitstbl meta data and meta data from spectrograph
+        and meta data from original header
+
         Args:
-            head_fitstbl (header):
+            head_fitstbl (:class:`astropy.io.fits.Header`):
```
Comment (kbwestfall) [inline pypeit/specobjs.py]: Just a heads-up: For stuff that's outside pypeit, all the `:class:` prefix does is set the font for the class path.  For this kind of thing, I tend to do:
```
head_fitstbl (`astropy.io.fits.Header`_)
```
where the type is now a link defined in `pypeit/doc/links.rst`.  At the top of the file you'll see the lines:
```
.. include:: ../links.rst
```
which includes the link definitions for all docstrings in the module.

File: pypeit/specobjs.py [hunk 9]
```diff
@@ -506,23 +517,35 @@ def __repr__(self):
...
+               Header from the Raw file
+            spectrograph (:class:`pypeit.spectrographs.spectrograph.Spectrograph`):
                Spectrograph object
 
         Returns:
+            :class:`astropy.io.fits.Header`:
 
         """
-        header = fits.PrimaryHDU().header
+        header = initialize_header(primary=True)
```
Comment (kbwestfall) [inline pypeit/specobjs.py]: I don't think we need the `primary` keyword, but am happy to be proven wrong.

File: pypeit/tests/test_flux.py [hunk 10]
```diff
@@ -109,3 +110,15 @@ def test_extinction_correction():
...
+    # Test scale_in_filter() method which is called in coadding
+    wave = np.arange(3000.,10000.)
+    flux = np.ones_like(wave)
+    gdm = np.ones_like(wave, dtype=bool)
+    #
+    par = Coadd1DPar()
+    par['filter'] = 'DECAM-R'
+    par['filter_mag'] = 17.
+    # Run
+    scale = flux_calib.scale_in_filter(wave, flux, gdm, par)
```
Comment (kbwestfall) [inline pypeit/tests/test_flux.py]: Does this need to take the full parameter set? I.e., can we just pass in `filter` and `filter_mag`?

File: CHANGES.rst [hunk 11]
```diff
@@ -18,6 +18,12 @@
   key is not valid for the given parameter set.  NOTE: The check may
   fail if there are identical keys for different parameter sets.
 - Modification to add_sobj() for numpy 18
+- Enable user to scale flux of coadded 1D spectrum to a filter magnitude
```
[the remark this answers is not in the document]
Comment (profxj) [reply CHANGES.rst]: done

File: doc/installing.rst [hunk 12]
```diff
@@ -31,7 +31,7 @@ We recommend that you use `Anaconda <https://www.continuum.io/downloads/>`_ to i
 * `numba <https://numba.pydata.org/>`_ version 0.39.0 or later (optional - some speed ups, inc. wavecal)
 * `PyQT5 <https://wiki.python.org/moin/PyQt/>`_ version 5 (needed for linetools)
 * `h5py <https://www.h5py.org/>`_ version 2.7 (for data I/O)
-* yaml -- You may need to install pyyaml
+* yaml -- version 5.1
```
[the remark this answers is not in the document]
Comment (profxj) [reply doc/installing.rst]: done

File: pypeit/core/flux_calib.py [hunk 13]
```diff
@@ -139,6 +133,10 @@ def find_standard_file(ra, dec, toler=20.*units.arcmin, check=False):
                 std_dict['wave'] = std_spec['col1'] * units.AA
                 std_dict['flux'] = std_spec['col2']*1e-16/PYPEIT_FLUX_SCALE * \
                                    units.erg / units.s / units.cm ** 2 / units.AA
+                # At this low resolution, best to throw out entries affected by A and B-band absorption
+                mask = (std_dict['wave'].value > 7551.) & (std_dict['wave'].value < 7749.)
+                std_dict['wave'] = std_dict['wave'][np.invert(mask)]
+                std_dict['flux'] = std_dict['flux'][np.invert(mask)]
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/flux_calib.py]: Fixed.

I believe it is just packaging.  It does get a bit unwieldy (additional items)
are added downstream.

File: pypeit/core/flux_calib.py [hunk 14]
```diff
@@ -673,8 +672,8 @@ def sensfunc_eval(wave, counts, counts_ivar, counts_mask, exptime, airmass, std_
         polycorrect=polycorrect, debug=debug, show_QA=False)
 
     if debug:
-        plt.plot(wave_star.value[mask_sens], flux_true[mask_sens], color='k',lw=2,label='Reference Star')
-        plt.plot(wave_star.value[mask_sens], flux_star[mask_sens]*sensfunc[mask_sens], color='r',label='Fluxed Observed Star')
+        plt.plot(wave_star[mask_sens], flux_true[mask_sens], color='k',lw=2,label='Reference Star')
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/flux_calib.py]: done

File: pypeit/scripts/show_2dspec.py (hunk 6 above)
Comment (profxj) [reply pypeit/scripts/show_2dspec.py]: good eye.  Am just commenting out for now.

File: pypeit/sensfunc.py (hunk 7 above)
Comment (profxj) [reply pypeit/sensfunc.py]: removed the `figure`.  Am pretty sure I need the `close`

File: pypeit/specobjs.py (hunk 9 above)
Comment (profxj) [reply pypeit/specobjs.py]: In this case we do.  `initialize_header()` will not 
by default make a `PrimaryHDU().header`.

This was biting me somewhere..

File: pypeit/tests/test_flux.py (hunk 10 above)
Comment (profxj) [reply pypeit/tests/test_flux.py]: we need `mag_type` too and `filter_mask`.
I drew the line at 4.  ;)

File: pypeit/core/flux_calib.py [hunk 15]
```diff
@@ -1076,41 +1076,42 @@ def scale_in_filter(xspec, scale_dict):
...
+    if scale_dict['filter_mask'] is not None:
+        # Funny formatting
+        if isinstance(scale_dict['filter_mask'], str):
+            regions = scale_dict['filter_mask'].split(',')
+        else:
+            regions = scale_dict['filter_mask']
+        for region in regions:
+            mask = region.split(':')
+            bad = (wave > float(mask[0])) & (wave < float(mask[1]))
+            gdm[bad] = False
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/flux_calib.py]: modified.

and edited

File: pypeit/core/wavecal/templates.py (hunk 1 above)
Comment (profxj) [reply pypeit/core/wavecal/templates.py]: Good suggstion. And I will save it for a later day.

File: pypeit/io.py (hunk 2 above)
Comment (profxj) [reply pypeit/io.py]: I have to admit my memory is fuzzy.  I think it was
in a corner case of re-running pypeit (e.g. when 
we are adding to an existing `spec1d` file.).

File: pypeit/metadata.py (hunk 3 above)
Comment (profxj) [reply pypeit/metadata.py]: agreed

Comment (profxj) [reply pypeit/metadata.py]: fixed.

File: pypeit/metadata.py (hunk 4 above)
Comment (profxj) [reply pypeit/metadata.py]: Was probably debugging via a test.
Edited.

File: pypeit/metadata.py (hunk 5 above)
Comment (profxj) [reply pypeit/metadata.py]: much cleaner!  Thanks.

**Comment by profxj** [comment]
--- PYPEIT DEVELOPMENT SUITE FAILED 4/57 TESTS (Masters ignored) ---
Failed tests:
    pypeit_sensfunc: gemini_gmos/GS_HAM_R400_860
    pypeit_coadd_1dspec: gemini_gmos/GS_HAM_R400_860
    run_pypeit: lbt_luci/LUCI-I
    run_pypeit: lbt_luci/LUCI-II

Working to resolve those..

**Comment by profxj** [comment]
All now fixed.  One requires the `gmos_fluxing` branch in the Dev suite repo.

File: pypeit/io.py (hunk 2 above)
Comment (profxj) [reply pypeit/io.py]: considered resolved

File: pypeit/core/coadd.py [hunk 16]
```diff
@@ -3073,6 +3074,11 @@ def run(self):
         self.waves, self.fluxes, self.ivars, self.masks, self.header = self.load()
         # Coadd the data
         self.wave_coadd, self.flux_coadd, self.ivar_coadd, self.mask_coadd = self.coadd()
+        # Scale to a filter magnitude?
```
Comment (jhennawi) [inline pypeit/core/coadd.py]: Should this be a command line argument rather than a parset parameter. The problem is we have only one parser for the .coadd files, and we tend to lump everything into parsets, but I'd much rather separate defaults and parameters which are in the parset from run time parameters which should be passed via the command line or passed via the .coadd, .flux etc. files.

File: pypeit/core/flux_calib.py [hunk 17]
```diff
@@ -56,27 +54,23 @@ def find_standard_file(ra, dec, toler=20.*units.arcmin, check=False):
...
     # Priority
     std_sets = ['xshooter', 'calspec', 'esofil']
 
     # SkyCoord
-    try:
-        ra, dec = float(ra), float(dec)
-        obj_coord = coordinates.SkyCoord(ra, dec, unit=(units.deg, units.deg))
-    except:
-        obj_coord = coordinates.SkyCoord(ra, dec, unit=(units.hourangle, units.deg))
+    obj_coord = coordinates.SkyCoord(ra, dec, unit='deg')
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py]: Is there a reason why we are removing support for multiple formats? Did something change upstream?

File: pypeit/core/flux_calib.py [hunk 18]
```diff
@@ -139,6 +133,10 @@ def find_standard_file(ra, dec, toler=20.*units.arcmin, check=False):
                 std_dict['wave'] = std_spec['col1'] * units.AA
                 std_dict['flux'] = std_spec['col2']*1e-16/PYPEIT_FLUX_SCALE * \
                                    units.erg / units.s / units.cm ** 2 / units.AA
+                # At this low resolution, best to throw out entries affected by A and B-band absorption
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py]: I'm not following why you are masking the standard here, whereas we have a routine that basically masks Balmer lines, etc.

File: pypeit/metadata.py [hunk 19]
```diff
@@ -114,7 +114,7 @@ def __init__(self, spectrograph, par, files=None, data=None, usrdata=None, stric
                                  else self._build(files, strict=strict, usrdata=usrdata))
 
         # Sort on filename
-        self.table.sort('filename')
+        #self.table.sort('filename')  # JXP -- This was causing issues with merging, and I think we sort on MJD later
```
Comment (jhennawi) [inline pypeit/metadata.py]: This is fine with me as long as the output pypeit file is sorted by MJD.

File: pypeit/par/pypeitpar.py [hunk 20]
```diff
@@ -686,7 +686,8 @@ class Coadd1DPar(ParSet):
     def __init__(self, ex_value=None, flux_value=None, nmaskedge=None,
                  sn_smooth_npix=None, wave_method=None, samp_fact=None, ref_percentile=None, maxiter_scale=None,
                  sigrej_scale=None, scale_method=None, sn_min_medscale=None, sn_min_polyscale=None, maxiter_reject=None,
-                 lower=None, upper=None, maxrej=None, sn_clip=None, nbest=None, sensfuncfile=None, coaddfile=None):
+                 lower=None, upper=None, maxrej=None, sn_clip=None, nbest=None, sensfuncfile=None, coaddfile=None,
+                 mag_type=None, filter=None, filter_mag=None, filter_mask=None):
```
Comment (jhennawi) [inline pypeit/par/pypeitpar.py]: To me the mags for renormalizing should be parameters not passed in via parsets, but I think this would require another way to think about file parsing. I will tolerate it for now.

File: pypeit/par/pypeitpar.py [hunk 21]
```diff
@@ -807,6 +808,24 @@ def __init__(self, ex_value=None, flux_value=None, nmaskedge=None,
...
+
+        defaults['filter_mag'] = None
+        dtypes['filter_mag'] = float
+        descr['filter_mag'] = 'Magnitude of the source in the given filter'
+
+        defaults['filter_mask'] = None
+        dtypes['filter_mask'] = [str, list]
+        descr['filter_mask'] = 'List of wavelength regions to mask when doing the scaling (ie. occasional junk pixels).'\
+                               'Colon and comma separateed, e.g.   5552:5559,6010:6030'
+
```
Comment (jhennawi) [inline pypeit/par/pypeitpar.py]: This is my point about the arguments versus parameters.

File: pypeit/scripts/show_2dspec.py [hunk 22]
```diff
@@ -133,7 +133,15 @@ def main(args):
     slit_ids = [edgetrace.get_slitid(shape, tslits_dict['slit_left'], tslits_dict['slit_righ'],
                                      ii)[0] for ii in range(tslits_dict['slit_left'].shape[1])]
     # Show the bitmask?
-    mask_in = mask if args.showmask else None
+    if args.showmask:
```
Comment (jhennawi) [inline pypeit/scripts/show_2dspec.py]: Why are you importing the ImageBitMask here and then passing it in. You could just do this in show_image, and apply the bitmask if the mask_in is passed. Not sure why you need to pass this instance here rather than just creating it in show_image. Is the issue that you just want to create it once?

File: pypeit/sensfunc.py [hunk 23]
```diff
@@ -255,12 +261,24 @@ def splice(self, wave):
             splice_wave_mask = (wave_splice >= wave_mask_min) & (wave_splice <= wave_mask_max)
             sensfunc_splice[splice_wave_mask] = self.eval_sensfunc(wave_splice[splice_wave_mask], idet)
 
+        # Interpolate over gaps
```
Comment (jhennawi) [inline pypeit/sensfunc.py]: Is this here to deal with cases where the standard has gaps in a different location than the science data (because the grating was tilted)? It is a bit of a weird thing to do in the limit where the detectors have different throughput, and it is obviously incorrect in that regime.

File: pypeit/specobjs.py [hunk 24]
```diff
@@ -241,14 +241,19 @@ def get_std(self, multi_spec_det=None):
...
-            SpecObj or SpecObjs
+            SpecObj or SpecObjs or None
 
         """
         # Is this MultiSlit or Echelle
         pypeline = (self.PYPELINE)[0]
         if 'MultiSlit' in pypeline:
             # Have to do a loop to extract the counts for all objects
-            SNR = np.median(self.OPT_COUNTS*np.sqrt(self.OPT_COUNTS_IVAR), axis=1)
+            if hasattr(self, 'OPT_COUNTS'):
```
Comment (jhennawi) [inline pypeit/specobjs.py]: A pedantic point, but to me the cleaner logic is:

if OPT_COUNTS
elif BOX_COUNTS
else None.

File: pypeit/specobjs.py [hunk 25]
```diff
@@ -274,7 +279,13 @@ def get_std(self, multi_spec_det=None):
                 for iord in range(norders):
                     ind = (self.ECH_FRACPOS == uni_objid[iobj]) & (self.ECH_ORDER == uni_order[iord])
                     spec = self[ind]
-                    SNR[iord, iobj] = np.median(spec[0].OPT_COUNTS*np.sqrt(spec[0].OPT_COUNTS_IVAR))
+                    # Grab SNR
+                    if hasattr(spec[0], 'OPT_COUNTS'):
```
Comment (jhennawi) [inline pypeit/specobjs.py]: Same comment above applies here.

File: pypeit/spectrographs/gemini_gmos.py [hunk 26]
```diff
@@ -400,10 +399,10 @@ def bpm(self, filename, det, shape=None, msbias=None):
             # Apply the mask
             xbin = int(binning.split(' ')[0])
             if xbin != 2:
```
Comment (jhennawi) [inline pypeit/spectrographs/gemini_gmos.py]: Perhaps the code should fault rather than embed.

File: pypeit/spectrographs/spectrograph.py [hunk 27]
```diff
@@ -573,8 +578,15 @@ def get_meta_value(self, inp, meta_key, required=False, ignore_bad_header=False,
                 value = headarr[self.meta[meta_key]['ext']][self.meta[meta_key]['card']]
             except (KeyError, TypeError):
                 value = None
+        # Return now?
```
Comment (jhennawi) [inline pypeit/spectrographs/spectrograph.py]: Can you explain why we need this option?

File: pypeit/pypeit.py [hunk 28]
```diff
@@ -481,6 +480,9 @@ def get_std_trace(self, std_redux, det, std_outfile):
             if np.any(this_det):
                 sobjs_det = sobjs[this_det]
                 sobjs_std = sobjs_det.get_std()
+                # No standard extracted??
```
Comment (jhennawi) [inline pypeit/pypeit.py]: Please change this comment to 

#No standard extracted on this detector

File: pypeit/core/coadd.py (hunk 16 above)
Comment (profxj) [reply pypeit/core/coadd.py]: Well, it would require 4 command line arguments.
And it is nice having them recorded in the coadd file.

File: pypeit/core/flux_calib.py (hunk 18 above)
Comment (profxj) [reply pypeit/core/flux_calib.py]: The default masking was working poorly for these ~50A binned ESO standards.

File: pypeit/metadata.py (hunk 19 above)
Comment (profxj) [reply pypeit/metadata.py]: The MJD sort has not been removed.

File: pypeit/par/pypeitpar.py (hunk 20 above)
Comment (profxj) [reply pypeit/par/pypeitpar.py]: Yup.

File: pypeit/scripts/show_2dspec.py (hunk 22 above)
Comment (profxj) [reply pypeit/scripts/show_2dspec.py]: I had tried to use this functionality.  
Then failed. Am pretty sure this 
preserves the original intent.

File: pypeit/sensfunc.py (hunk 23 above)
Comment (profxj) [reply pypeit/sensfunc.py]: Am pretty sure it is the gaps in the GMOS 3 detectors.

File: pypeit/specobjs.py (hunk 24 above)
Comment (profxj) [reply pypeit/specobjs.py]: Agreed.  Modified.

File: pypeit/specobjs.py (hunk 25 above)
Comment (profxj) [reply pypeit/specobjs.py]: Also agreed. And modified.
And changed `self` to `spec[0]`

File: pypeit/spectrographs/gemini_gmos.py (hunk 26 above)
Comment (profxj) [reply pypeit/spectrographs/gemini_gmos.py]: modified.

File: pypeit/spectrographs/spectrograph.py (hunk 27 above)
Comment (profxj) [reply pypeit/spectrographs/spectrograph.py]: It's not elegant.  But if you look a few lines down, you see 
the calls to `get_meta_value()` for `ra,dec` that use this. 

Without this option, the crash gets into an infinite loop...

File: pypeit/pypeit.py (hunk 28 above)
Comment (profxj) [reply pypeit/pypeit.py]: Done

File: pypeit/sensfunc.py (hunk 23 above)
Comment (jhennawi) [reply pypeit/sensfunc.py]: That is clear, but for a single setup you never have data in those gaps, so no need to interpolate. For multiple setups where you dither in wavelength, the standard Gemini approach, you are doing something weird, because data in the gap is covered by two different detectors.

**Comment by profxj** [comment]
--- PYPEIT DEVELOPMENT SUITE FAILED 1/57 TESTS (Masters ignored) ---
Failed tests:
    pypeit_ql: keck_nires/NIRES

**Comment by profxj** [comment]
Was missing NIRES_MASTERS.  Passed:

[INFO]    :: pypeit.py 316 reduce_all() - Finished calibration group 0
[INFO]    :: pypeit.py 640 print_end_time() - Execution time: 1m 18.30s
[INFO]    :: ql_keck_nires.py 93 main() - Data reduction complete
[INFO]    :: ql_keck_nires.py 95 main() - Generating QA HTML
Wrote: /data/Projects/Python/PypeIt-development-suite/REDUX_OUT/keck_nires/QL/keck_nires_A/QA/MF_A.html


Ok @jhennawi -- Time to merge.

**Comment by profxj** [comment]
Merging!
